source/common/stream_info/utility.cc

```diff
@@ -43,6 +47,230 @@ absl::optional<ResponseFlag> ResponseFlagUtils::toResponseFlag(absl::string_view
+  case ProxyStatusError::DestinationUnavailable:
+    return DESTINATION_UNAVAILABLE;
+  case ProxyStatusError::DestinationIpProhibited:
+    return DESTINATION_IP_PROHIBITED;
+  case ProxyStatusError::DestinationIpUnroutable:
+    return DESTINATION_IP_UNROUTABLE;
+  case ProxyStatusError::ConnectionRefused:
+    return CONNECTION_REFUSED;
+  case ProxyStatusError::ConnectionTerminated:
+    return CONNECTION_TERMINATED;
+  case ProxyStatusError::ConnectionTimeout:
+    return CONNECTION_TIMEOUT;
+  case ProxyStatusError::ConnectionReadTimeout:
+    return CONNECTION_READ_TIMEOUT;
+  case ProxyStatusError::ConnectionWriteTimeout:
+    return CONNECTION_WRITE_TIMEOUT;
+  case ProxyStatusError::ConnectionLimitReached:
+    return CONNECTION_LIMIT_REACHED;
+  case ProxyStatusError::TlsProtocolError:
+    return TLS_PROTOCOL_ERORR;
+  case ProxyStatusError::TlsCertificateError:
+    return TLS_CERTIFICATE_ERORR;
+  case ProxyStatusError::TlsAlertReceived:
+    return TLS_ALERT_RECEIVED;
+  case ProxyStatusError::HttpRequestError:
+    return HTTP_REQUEST_ERROR;
+  case ProxyStatusError::HttpRequestDenied:
+    return HTTP_REQUEST_DENIED;
+  case ProxyStatusError::HttpResponseIncomplete:
+    return HTTP_RESPONSE_INCOMPLETE;
+  case ProxyStatusError::HttpResponseHeaderSectionSize:
+    return HTTP_RESPONSE_HEADER_SECTION_SIZE;
+  case ProxyStatusError::HttpResponseHeaderSize:
+    return HTTP_RESPONSE_HEADER_SIZE;
+  case ProxyStatusError::HttpResponseBodySize:
+    return HTTP_RESPONSE_BODY_SIZE;
+  case ProxyStatusError::HttpResponseTrailerSectionSize:
+    return HTTP_RESPONSE_TRAILER_SECTION_SIZE;
+  case ProxyStatusError::HttpResponseTrailerSize:
+    return HTTP_RESPONSE_TRAILER_SIZE;
+  case ProxyStatusError::HttpResponseTransferCoding:
+    return HTTP_RESPONSE_TRANSFER_CODING;
+  case ProxyStatusError::HttpResponseContentCoding:
+    return HTTP_RESPONSE_CONTENT_CODING;
+  case ProxyStatusError::HttpResponseTimeout:
+    return HTTP_RESPONSE_TIMEOUT;
+  case ProxyStatusError::HttpUpgradeFailed:
+    return HTTP_UPGRADE_FAILED;
+  case ProxyStatusError::HttpProtocolError:
+    return HTTP_PROTOCOL_ERROR;
+  case ProxyStatusError::ProxyInternalResponse:
+    return PROXY_INTERNAL_RESPONSE;
+  case ProxyStatusError::ProxyInternalError:
+    return PROXY_INTERNAL_ERROR;
+  case ProxyStatusError::ProxyConfigurationError:
+    return PROXY_CONFIGURATION_ERROR;
+  case ProxyStatusError::ProxyLoopDetected:
+    return PROXY_LOOP_DETECTED;
+  }
+}
+
+const absl::optional<ProxyStatusError>
+ProxyStatusUtils::fromStreamInfo(const StreamInfo& stream_info) {
+  // NB: This mapping from Envoy-specific ResponseFlag enum to Proxy-Status
+  // error enum is lossy, since ResponseFlag is really a bitset of many
+  // ResponseFlag enums. Here, we search the list of all known ResponseFlag values in
+  // enum order, returning the first matching ProxyStatusError.
+  if (stream_info.hasResponseFlag(ResponseFlag::FailedLocalHealthCheck)) {
```

Comment:
I have concerns about a bunch of these.
I think a lot of these errors aren't mapped correctly, things like retry limit being exceeded being mapped to connection termination, and a general conflation of stream and connection.

I wonder if it'd make more sense to send specific proxy status errors as we do response details, where errors occur.  I worry that incorrect/misleading data is going to be worse than no data at all.  cc @mattklein123 @snowp for thoughts

Reply:
I don't have especially strong opinions about these mappings and I'm happy to change them based on feedback from you / anyone. I expected that this part of the PR would experience the most review. What specifically should I change here?

Reply:
We definitely loose some resolution here, but we can include the original flags in the `details` parameter